src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java

```diff
@@ -107,54 +115,95 @@
-          // where the due date field is required. Setting it to null is probably wrong and will
-          // likely break the template when building the SIP2 response.
-          if (resource.getResource() != null) {
-            final String dueDateString = resource.getResource().getString("dueDate", null);
-            if (dueDateString == null) {
-              dueDate = null;
-            } else {
-              // Need to convert to the tenant local timezone
-              dueDate = OffsetDateTime.from(Utils.getFolioDateTimeFormatter().parse(dueDateString));
-            }
+    return loginFuture
+        .compose(login -> {
+          final JsonObject body = new JsonObject()
+              .put("itemBarcode", itemIdentifier)
+              .put("userBarcode", lmsPatronIdentifier[0])
+              .put("servicePointId", sessionData.getScLocation());
+
+          final Map<String, String> headers = getBaseHeaders();
+
+          final CheckoutRequestData checkoutRequestData =
+              new CheckoutRequestData(body, headers, sessionData);
+
+          // This means we tried to log in as the patron and failed in some way
+          if (login != null && !login.getErrorMessages().isEmpty()) {
+            return Future.succeededFuture(CheckoutResponse.builder()
+                .ok(FALSE)
+                .renewalOk(FALSE)
+                .magneticMedia(null)
+                .desensitize(FALSE)
+                .transactionDate(OffsetDateTime.now(clock))
+                .institutionId(institutionId)
+                .patronIdentifier(patronIdentifier)
+                .itemIdentifier(itemIdentifier)
+                .titleIdentifier(UNKNOWN)
+                .dueDate(OffsetDateTime.now(clock))
+                .screenMessage(login.getErrorMessages())
+                .build());
           } else {
-            dueDate = null;
+            final Future<IResource> result = resourceProvider.createResource(checkoutRequestData);
+
+            return result
+                .otherwise(Utils::handleErrors)
+                .map(resource -> {
+                  final Optional<JsonObject> response = Optional.ofNullable(resource.getResource());
+                  // The problem here is that the checkout has already succeeded, but something
+                  // could be wrong with the returned data. The odds of this are low. The "dueDate"
+                  // field is not required, so a loan could be missing one. I am not sure
+                  // what that means for SIP2 where the due date field is required. Setting it to
+                  // null is probably wrong and will likely break the template when building the
+                  // SIP2 response.
+                  final OffsetDateTime dueDate = response
+                      .map(v -> v.getString("dueDate"))
+                      .map(v -> OffsetDateTime.from(Utils.getFolioDateTimeFormatter().parse(v)))
+                      .orElse(null);
```

Comment:
can we give it an empty string instead? Henceforth empty string for due dates signifies a problem? May want to add a log statement here explaining what could have happened when duedate is null or empty string.

Reply:
According to Marc, `dueDate` should never be `null`. Maybe we should just assume it will always be set and take out the `orElse`.

> Marc, another question, when you get a chance. What does it mean if a loan is missing the `dueDate` field? Does that mean that the loan is never overdue? For now, I am going to assume `dueDate` will always be present, even though it is not a required field in the schema.

> An open loan shouldn’t ever not have a due date
> In general, a loan shouldn’t not have a due date, adding it in would be a breaking change

Reply:
Yeah, let's assume so and log it (or add a count metric if easily doable) to track if it ever being null?

Reply:
Actually, reviewing the code again, the `dueDate` can be null if checkout has failed on the FOLIO side. So, now I think we should assume that `dueDate` will be there on success and when it is not there, it is a failure case, and the `dueDate` is irrelevant, but needs to be set to something, so we'll set it to the current date.

```suggestion
                      .orElse(OffsetDateTime.now(clock));
```

@mtraneis does that work your you? We won't be able to track (unless we put in more logic) the case that should never happen where the FOLIO checkout response didn't have a due date.